osu.Server.Spectator/GracefulShutdownManager.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using osu.Framework.Logging;
+using osu.Server.Spectator.Entities;
+
+namespace osu.Server.Spectator;
+
+/// <summary>
+/// Ensures that shutdown is delayed until any existing usages have ceased.
+/// </summary>
+public class GracefulShutdownManager
+{
+    public static readonly TimeSpan TIME_BEFORE_FORCEFUL_SHUTDOWN = TimeSpan.FromHours(6);
+
+    private readonly List<IEntityStore> dependentStores = new List<IEntityStore>();
+
+    private Task? shutdownTask;
+
+    /// <summary>
+    /// Blocks until safe to continue with shutdown. Can be invoked from multiple locations.
+    /// </summary>
+    public void WaitForSafeShutdown()
+    {
+        lock (this)

```

Comment:
Was `lock (this)` intentional in this context? Usually this type of thing is done on a dummy private variable (`private object syncRoot = new object();` or similar), to avoid external code from taking the lock on this instance and potentially causing deadlocks.

Reply:
It was intentional because nothing else is going to be locking this component. I can change it if required.

Reply:
Has been removed as no longer required.